src/options.js

```diff
@@ -3,21 +3,43 @@ import {color, descending, range as rangei, quantile} from "d3";
 import {maybeTimeInterval, maybeUtcInterval} from "./time.js";
 
 // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray
-const TypedArray = Object.getPrototypeOf(Uint8Array);
+export const TypedArray = Object.getPrototypeOf(Uint8Array);
 const objectToString = Object.prototype.toString;
 
 /** @jsdoc valueof */
 export function valueof(data, value, type) {
   const valueType = typeof value;
   return valueType === "string"
-    ? map(data, field(value), type)
+    ? maybeTypedMap(data, field(value), type)
     : valueType === "function"
-    ? map(data, value, type)
+    ? maybeTypedMap(data, value, type)
     : valueType === "number" || value instanceof Date || valueType === "boolean"
     ? map(data, constant(value), type)
     : value && typeof value.transform === "function"
-    ? arrayify(value.transform(data), type)
-    : arrayify(value, type); // preserve undefined type
+    ? maybedTypedArrayify(value.transform(data), type)
+    : maybedTypedArrayify(value, type);
+}
+
+// When valueof is asked to produce a typed array (i.e., numbers) we implicitly
+// apply null-safe type coercion.
+function maybeTypedMap(data, f, type) {
+  return map(data, type?.prototype instanceof TypedArray ? floater(f) : f, type);
+}
+
+// When valueof is asked to produce a typed array (i.e., numbers) we implicitly
+// apply null-safe type coercion.
+function maybedTypedArrayify(data, type) {
+  return type === undefined
+    ? arrayify(data) // preserve undefined type
+    : data instanceof type
+    ? data
+    : type.prototype instanceof TypedArray && !(data instanceof TypedArray)
```

Comment:
Bonus: if we already have a typed array, we can skip the coercion (even if we have to change the array type)!